### PR TITLE
Configure logging via TOML config

### PR DESCRIPTION
**Pull request by orsinium**
## Description

1. Now you can fully configure logging from TOML-config.
2. Option `no_configure_logging` disable any logging configuring.
3. Separate logging configuration from other code for better maintainability.
4. Also I update Apache copyright notice from my previous PR. Cindicator's leader recommend to use "Vote Inc." company name in copyrights.

## Motivation and Context

Q. Why do you use same file for Luigi's config and logging config?
A. Because this is all project config. Also in future Pull Requests I plan to integrate in Luigi configs inheritance which allow also use separate file for logging, local settings or something else.

Q. Why do you do this? We already have support for `logging.conf`.
A. Because logging via ini-file is ugly, since ini format doesn't support inherit dicts. But TOML (and YAML. JSON and other modern formats) can do it.

Q. Why do you change `no_configure_logging` behavior?
A. Option `no_configure_logging` useful when you're configure logging in project bypass to Luigi. For example, when you want to use [JSON or YAML configs](https://fangpenlin.com/posts/2012/08/26/good-logging-practice-in-python/) for it. This is intuitive behavior for option with this name.

Q. Why do you change code structure?
A. Old code structure is hard to test, explore and extend. Now it's beautiful :)

## Have you tested this? If so, how?

1. I have included unit tests.
2. Don't trust me! I update [example repository](https://github.com/orsinium/luigi-example), and you can play with new logging configuring yourself.

## And...

Thank you for best pipeline library. And best music streaming service too. All this things is amazing!

### CODE REVIEW

**Comment by orsinium** [comment]
Sorry for many commits. This is going from my previous PR, because you squash all my changes instead of merge.

File: doc/logging.rst [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+``--logging-conf-file``
+    Configuration file for logging.
+``--log-level``
+    Default log level.
+    Available values: NOTSET, DEBUG, INFO, WARNING, ERROR, CRITICAL.
+    Default DEBUG.
+
+
+Ways priority:
```
[the remark this answers is not in the document]
Comment (orsinium) [reply doc/logging.rst]: Logging resolution order

File: doc/logging.rst [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+~~~~~~~~~~~~~~
+
+1. no_configure_logging option
+1. ``--background``
+1. ``--logdir``
+1. ``--logging-conf-file``
+1. logging_conf_file option
+1. ``logging`` section
+1. ``--log-level``
+1. log_level option
```
[the remark this answers is not in the document]
Comment (orsinium) [reply doc/logging.rst]: Sorry. I test it in pandoc, it is automatically setup right numbers like in markdown. I'll fix it

**Comment by orsinium** [comment]
I have some failed tests. I set up WIP status for project while I not find source of this troubles.

> Oh and could you update the PR Title to be more clear?

You're right, fixed.

File: doc/logging.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (dlstadther) [reply doc/logging.rst]: Ah, that's more clear. Please update documentation language :)

File: doc/logging.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (dlstadther) [reply doc/logging.rst]: I'm not overly proficient with rst. So i'm not sure if/how to auto number.

[the remark this answers is not in the document]
Comment (orsinium) [reply doc/logging.rst]: Fixed

File: doc/logging.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (orsinium) [reply doc/logging.rst]: Fixed

File: doc/logging.rst [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+
+Worker CLI options:
+~~~~~~~~~~~~~~~~~~~
+
+``--logging-conf-file``
+    Configuration file for logging.
+``--log-level``
+    Default log level.
+    Available values: NOTSET, DEBUG, INFO, WARNING, ERROR, CRITICAL.
```
[the remark this answers is not in the document]
Comment (orsinium) [reply doc/logging.rst]: Fixed

File: doc/logging.rst [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+Config options:
+~~~~~~~~~~~~~~~
+
+Some config options for config [core] section
+
+log_level
+    The default log level to use when no logging_conf_file is set. Must be
+    a valid name of a `Python log level
+    <https://docs.python.org/2/library/logging.html#logging-levels>`_.
```
[the remark this answers is not in the document]
Comment (orsinium) [reply doc/logging.rst]: Fixed

File: luigi/setup_logging.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        If `--logdir` -- set logging with next params:
+            default Luigi's formatter,
+            INFO level,
+            output in logdir in `luigi-server.log` file
+        """
+        if opts.background:
+            logging.getLogger().setLevel(logging.INFO)
+            return True
+
+        if opts.logdir:
```
[the remark this answers is not in the document]
Comment (orsinium) [reply luigi/setup_logging.py]: Can we swap order of background and logdir? What do you think?

**Comment by Tarrasch** [comment]
I'm in favor of this. I'll have to review another time though. :)

File: luigi/setup_logging.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dlstadther) [reply luigi/setup_logging.py]: As far as I can tell from the history of Luigi, `opts.background` has always taken the priority. I'm fine with leaving it this way so long as we note in documentation somewhere.

[the remark this answers is not in the document]
Comment (orsinium) [reply luigi/setup_logging.py]: Now it's documented :)

**Comment by orsinium** [comment]
@Tarrasch, how is your review? ^^

File: test/setup_logging_test.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        self.assertTrue(result)
+
+    def test_default(self):
+        opts = type('opts', (), {})
+        opts.log_level = 'INFO'
+        result = self.cls._default(opts)
+        self.assertTrue(result)
+
+
+class PatchedLogging(InterfaceLogging):
```
Comment (Tarrasch) [inline test/setup_logging_test.py]: Was it possible to mock one layer less?

File: test/cmdline_test.py [hunk 7]
```diff
@@ -135,35 +144,26 @@ def test_cmdline_ambiguous_class(self, logger):
...
-        with mock.patch("luigi.interface.setup_interface_logging.has_run", new=False):
-            if six.PY2:
-                error = ConfigParser.NoSectionError
-            else:
-                error = KeyError
-            self.assertRaises(error, luigi.interface.setup_interface_logging, '/blah')
-
-    @mock.patch("warnings.warn")
-    @mock.patch("luigi.interface.setup_interface_logging")
-    def test_cmdline_logger(self, setup_mock, warn):
```
Comment (Tarrasch) [inline test/cmdline_test.py]: It's so much code that it's hard to review it all. But can you make sure that all tests you delete are added somewhere else?

File: test/setup_logging_test.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        opts.logdir = './tests/'
+        result = self.cls._cli(opts)
+        self.assertTrue(result)
+
+        opts.background = False
+        opts.logdir = False
+        result = self.cls._cli(opts)
+        self.assertFalse(result)
+
+    def test_section(self):
```
Comment (Tarrasch) [inline test/setup_logging_test.py]: This test name `setup_logging_test.TestDaemonLogging.test_section` could have more context if you rename `test_section` to maybe `test_section_returns_true_with_config` and another test `test_section_returns_false_without_config`.

I know we don't follow this style at most places in this code base. But code testing improvements can start somewhere. If you want you can do this is too much you can do this as a follow up.

File: test/cmdline_test.py [hunk 9]
```diff
@@ -112,6 +108,19 @@ class CmdlineTest(unittest.TestCase):
...
         MockTarget.fs.clear()
+        DaemonLogging.configured = False
+
+    def tearDown(self):
+        DaemonLogging.configured = False
+        DaemonLogging.config = get_config()
+        InterfaceLogging.config = get_config()
+
+    def _clean_config(self):
+        DaemonLogging.config = LuigiTomlParser()
```
[the remark this answers is not in the document]
Comment (orsinium) [reply test/cmdline_test.py]: It is for creating minimal config for tests. It's here for old tests.

File: test/cmdline_test.py (hunk 7 above)
Comment (orsinium) [reply test/cmdline_test.py]: Yes, all logging well tested in `test/setup_logging_test.py`

File: test/setup_logging_test.py (hunk 8 above)
Comment (orsinium) [reply test/setup_logging_test.py]: It's test all what we need test for `_section` method

**Comment by orsinium** [comment]
Added some @Tarrasch's improvements. Other requests commented

File: test/setup_logging_test.py (hunk 8 above)
Comment (Tarrasch) [reply test/setup_logging_test.py]: The point is that `test_section` is not saying much compared to `test_section_returns_true_with_config`.

File: luigi/setup_logging.py [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+            # FileNotFoundError added only in Python 3
+            # https://docs.python.org/2/library/exceptions.html#exception-hierarchy
+            # https://docs.python.org/3/library/exceptions.html#exception-hierarchy
+            raise OSError("Error: Unable to locate specified logging configuration file!")
+
+        logging.config.fileConfig(logging_conf)
+        return True
+
+    @classmethod
+    def _default(cls, opts):
```
[the remark this answers is not in the document]
Comment (Tarrasch) [reply luigi/setup_logging.py]: Okay so you can stay consistent in this PR, but please consider following up with a PR with changes like this:

 * `_default` --> `_setup_default_logging`
 * `_conf` --> `_setup_ini_logging`

etc. What do you think?

**Comment by orsinium** [comment]
CI is OK now. What about this PR? :)

**Comment by dlstadther** [comment]
I'm good here.

Have you been using this feature in production? All is good for your use?

**Comment by orsinium** [comment]
Now I'm in another team in the same company. I'll get actual feedback on this feature by next few days :)

**Comment by dlstadther** [comment]
Congrats on your move @orsinium ! Appreciate the feedback

**Comment by orsinium** [comment]
I got feedback and have fixed example based on this:
1. `disable_existing_loggers` should be false for saving already existing loggers.
1. dot in `luigi.server` logging means inheritance so we can just configure only `luigi` logger.

I think, it will be better change `luigi-interface` to `luigi.interface`, but I don't do it in this PR because it can broke logging in existing projects.

So. now it is ready, I hope :)

**Comment by orsinium** [comment]
@dlstadther, @Tarrasch, what are we going to do with this Pull Request?

**Comment by dlstadther** [comment]
I just looked back through this PR and I think it LGTM. Given @Tarrasch had some changes requested, I'll wait for him to give his approval that his comments were addressed.

Thanks for continuing to follow up @orsinium !

File: luigi/interface.py [hunk 11]
```diff
@@ -166,17 +140,8 @@ def _schedule_and_run(tasks, worker_scheduler_factory=None, override_defaults=No
...
-    if logging_conf != '' and not os.path.exists(logging_conf):
-        raise Exception(
-            "Error: Unable to locate specified logging configuration file!"
-        )
-
-    if not configuration.get_config().getboolean(
-            'core', 'no_configure_logging', False):
-        setup_interface_logging(logging_conf, env_params.log_level)
+
+    InterfaceLogging.setup(env_params)
```
Comment (Tarrasch) [inline luigi/interface.py]: Oh nice you even cleaned up this part!

File: luigi/setup_logging.py [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""
+This module contains helper classes for configuring logging for luigid and
+workers via command line arguments and options from config files.
```
Comment (Tarrasch) [inline luigi/setup_logging.py]: Do these docs render nicely? Can you be explicit about if other's can use these methods or not?

File: luigi/setup_logging.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Tarrasch) [reply luigi/setup_logging.py]: (just a reminder to feel free to do this)

File: luigi/interface.py [hunk 13]
```diff
@@ -38,31 +36,7 @@
 from luigi import worker
 from luigi import execution_summary
 from luigi.cmdline_parser import CmdlineParser
-
-
-def setup_interface_logging(conf_file='', level_name='DEBUG'):
```
Comment (honnix) [inline luigi/interface.py]: A quick question. We call this method directly to set up logging. What is the equivalence now? Or it is not needed anymore?

**Comment by orsinium** [comment]
Yeah, of course, we call InterfaceLogging instead of:
https://github.com/spotify/luigi/blob/235edcccaa207b3c03c727521a95429cc736b568/luigi/interface.py#L144

**Pull request merged**